### PR TITLE
[VisBuilder] Adds UIState to vis, adds index patterns to embeddable, bug fixes

### LINKED ISSUE #1880: [Vis Builder] Complete unhappy save paths (error handling flows)
As a follow-up to #1870, we need to add additional handling of saving errors.

- [ ] [Use immer's `produce`](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/1870#discussion_r918451839) to [avoid mutating saved wizard](https://github.com/opensearch-project/OpenSearch-Dashboards/blob/8c50d0f87d4036e5bdb7bb18a09a3f52f214e6b3/src/plugins/wizard/public/application/utils/get_top_nav_config.tsx#L94-L98) on a failed save.
- [ ] Handle [validation failures](https://github.com/opensearch-project/OpenSearch-Dashboards/issues/1881)

### LINKED ISSUE #1881: [Vis Builder] Add validation steps between serialization/deserialization and state sync
As a follow-up to #1870

- [ ] Make sure states are valid [before serializing](https://github.com/opensearch-project/OpenSearch-Dashboards/blob/8c50d0f87d4036e5bdb7bb18a09a3f52f214e6b3/src/plugins/wizard/public/application/utils/get_top_nav_config.tsx#L94-L95), [handle errors](https://github.com/opensearch-project/OpenSearch-Dashboards/issues/1880)
- [x] Clean/remove unsupported properties [before dispatching state updates](https://github.com/opensearch-project/OpenSearch-Dashboards/blob/8c50d0f87d4036e5bdb7bb18a09a3f52f214e6b3/src/plugins/wizard/public/application/utils/use/use_saved_wizard_vis.ts#L47-L48)

### LINKED ISSUE #3124: [VisBuilder] Persist visualization internal state
VisBuilder Visualizations have internal state that they would like to persist when changed. Currently we do not persist this data. 

Note: We may want to hold off on implementing this until the migration to vega-lite is complete. f we cant wait for that, we need to investigate the impact of the migration on the saved object data since the uistate after migration may be drastically different from that of the legacy library.

### LINKED ISSUE #3429: [BUG] Can't get index-pattern fields at dashboard filter for vis-builder
**Describe the bug**

When a vis-builder visualization is added to dashboard, and user want to add some filter at dashboard top bar, the filter bar can’t get the fields from the index-patterns which the vis-builder used.

**To Reproduce**
1. Add all sample index-patterns at OSD
2. Create a new dashboard
3. Create new vis-builder from dashboard
4. In vis-builder creation page, select `opensearch_dashboards_sample_data_logs` index pattern and create a vis-builder vis
5. Saved and return back to dashboard
6. Add filter on dashboard top, select field
7. We can't see the fields options from `opensearch_dashboards_sample_data_logs`

**Expected behavior**
Get fields options at the filter bar for vis-builder index pattern source

**OpenSearch Version**
2.5.0 (OpenSearch playground env)
3.0.0 (Local dev env)

**Dashboards Version**
2.5.0 (OpenSearch playground env)
3.0.0 (Local dev env)

### LINKED ISSUE #3650: [Vis Builder] Support UiState in Vis Builder
**Describe the bug**
Currently Vis Builder does not support reading UiState, and this is blocking data table vis from being able to sort columns and persist the sorting columns. 

This is the related bug issue: https://github.com/opensearch-project/OpenSearch-Dashboards/issues/3642, and this issue occurs after 2.5 release when data table vis is replacing with React and DataGrid. Here is the PR: #2863

### PR DESCRIPTION
**Pull request by ashwin-pc**
> Note: This Change was built on top of #3732. Some the changes from that PR are in here. This is the diff for the change thats going in this PR: https://github.com/opensearch-project/OpenSearch-Dashboards/pull/3751/files/1472ee1ae26f19b4d8c49e37f65625af2f0076df..6fcace5b6038ae0a5dc41024369410df0c27f02a

### Description

This PR has a combination of changes and fixes as listed below:
1. Adds UI state to the saved object and metadata state to persist data
2. the multiple errors shown on page load even when there are no aggregations selected
3. Fixes the issue of being able to select the index pattern in the dashboard views filters when more than one index pattern exists (or if the index pattern differs from the default index pattern)
4. Fixes typescript errors
5. Simplifies the serializing and deserializing of saved object data with error handling

Demo:

https://user-images.githubusercontent.com/20453492/229004416-64094e3f-2d23-419a-b57d-2ae2be0dc366.mov

 
### Issues Resolved
- Closes #3650
- Closes #3124 
- Closes #1881
- Closes #1880 
- Fixes #3429 
 
### Check List
- [x] All tests pass
  - [ ] `yarn test:jest`
  - [ ] `yarn test:jest_integration`
  - [ ] `yarn test:ftr`
- [x] New functionality includes testing.
- [x] New functionality has been documented.
- [x] Update [CHANGELOG.md](./../CHANGELOG.md)
- [x] Commits are signed per the DCO using --signoff

### CODE REVIEW

**Comment by joshuarrrr** [comment]
@ashwin-pc When you rebase this, can you take a look at the remaining nits here? https://github.com/opensearch-project/OpenSearch-Dashboards/pull/3732#pullrequestreview-1386260487

**Comment by joshuarrrr** [comment]
@ashwin-pc CHANGELOG is missing.

File: src/plugins/vis_builder/public/saved_visualizations/_saved_vis.ts [hunk 1]
```diff
@@ -44,7 +45,8 @@ export function createSavedVisBuilderVisClass(services: SavedObjectOpenSearchDas
           description: '',
           visualizationState: '{}',
           styleState: '{}',
-          version: 2,
+          uiState: '{}',
+          version: 3,
```
Comment (joshuarrrr) [inline src/plugins/vis_builder/public/saved_visualizations/_saved_vis.ts:49]: Does this need a migration? To initially set the `uiState` for any previous VisBuilder objects?

File: src/plugins/vis_builder/public/application/utils/schema.json [hunk 2]
```diff
@@ -1,28 +1,47 @@
...
+    "uiState": {
+      "type": "object"
+    }
+  },
+  "required": [
+    "styleState",
+    "visualizationState"
+  ],
+  "additionalProperties": false
 }
```
Comment (joshuarrrr) [inline src/plugins/vis_builder/public/application/utils/schema.json:47]: nit - mission EOF newline

**Review by joshuarrrr — COMMENTED** [review]
@ashwin-pc no particular concerns. But it needs a CHANGELOG entry (or skip label), and the snapshots need updating.

File: src/plugins/vis_builder/public/application/utils/schema.json (hunk 2 above)
Comment (joshuarrrr) [reply src/plugins/vis_builder/public/application/utils/schema.json:47]: how did I type `mission` instead of `missing` 🤣

**Review by joshuarrrr — APPROVED** [review]

**Review by abbyhu2000 — APPROVED** [review]

**Review by joshuarrrr — DISMISSED** [review]

**Comment by joshuarrrr** [comment]
I'm waiting for automation co complete before resolving the CHANGELOG conflict.

**Comment by joshuarrrr** [comment]
All tests pass. I'll resolve the changelog conflict, and then merge without awaiting another test run.

**Review by joshuarrrr — APPROVED** [review]

**Pull request merged**